doc/neps/indexing.rst

```diff
@@ -0,0 +1,421 @@
+both new, and in many cases also old, users of NumPy. To avoid this problem
+and allow for more and clearer features, we propose (see below for
+explanations):
+
+1. Introduce ``arr.oindex[indices]`` which allows advanced indices, but
+   uses outer indexing logic.
+2. Introduce ``arr.vindex[indices]`` which use the current
+   "vectorized"/broadcasted logic but with two differences from
+   fancy indexing:
+       
+   1. Boolean indices always use the outer indexing logic.
+      (Multi dimensional booleans should be allowed).
+   2. The integer index result dimensions are always the first axes
+      of the result array. No transpose is done, even for a single
+      integer array index.
+
+3. Vanilla indexing on the array will only give warnings and eventually
+   errors either:
+     
+   * when there is ambiguity between legacy fancy and outer indexing
+     (note that ``arr[[1, 2], :, 0]`` is such a case, an integer
+     can be a "second" integer index array),
+   * when any integer index array is present (possibly additional for
+     more then one boolean index array).
+
+These constraints are sufficient for making indexing generally consistent
+with expectations.
+
+Note that all things mentioned here apply both for assignment as well as
+subscription.
+
+Understanding the details is not trivial. The `Examples` section gives code
+examples and the hopefully easier `Motivational Example` which provides some
+motivational use-cases for the general ideas.
+
+Motivation
+==========
+
+Old style advanced indexing with multiple array (boolean or integer) indices,
+also called "fancy indexing", tends to be very confusing for new users.
+While fancy (or legacy) indexing is useful in many cases one would naively
+assume that the result of multiple 1-d ranges is analogous to multiple
+slices along each dimension (also called "outer indexing").
+
+However, legacy fancy indexing with multiple arrays broadcasts these arrays
+into a single index over multiple dimensions. There are three main points
+of confusion when multiple array indices are involved:
+
+1. Most new users will usually expect outer indexing (consistent with
+   slicing). This is also the most common way of handling this in other
+   packages or languages.
+2. The axes introduced by the array indices are at the front, unless
+   all array indices are consecutive, in which case one can deduce where
+   the user "expects" them to be:
+
+   * `arr[:, [0, 1], :, [0, 1]]` will have the first dimension shaped 2.
+   * `arr[:, [0, 1], [0, 1]]` will have the second dimension shaped 2.
+
+3. When a boolean array index is mixed with another boolean or integer
+   array, the result is very hard to understand (the boolean array is
+   converted to integer array indices and then broadcast), and hardly
+   useful.
+   There is no well defined broadcast for booleans, so that boolean
+   indices are logically always "``outer``" type indices.
```

Comment:
Tick labels resulting from indexing are actually taken from the array being indexed rather than the array doing the indexing in xray and pandas. So indexing along multiple dimensions at once means that we need to collapse to something like a multi index / tuples. Likewise, it's really not obvious what to name the new axis name (I can think of several alternatives to your suggestion).

This doesn't mean we can't allow multidimensional indexing with oindex for numpy, but it is something unlikely to ever be supported by xray and pandas. Of course, it's quite straightforward for xray/pandas to simply raise an error in these situations. 

Does it ever make sense to do multidimensional indexing with anything other than a  Boolean array? If not, we should say so.